[Ee Kar Hee, Nicholas] iP

kheekheekhee opened the pull request:
# Duke

> "The inner machinations of my mind are an enigma." - Patrick Star ([source](https://youtu.be/1z-pVZiRjac))

Duke helps you keep track of tasks you have to do. It's
* text-based
* easy to learn
* ~~maybe slow :(~~ _EXTREMELY FAST_ 😫👌
* **FREE** 💦💦💦

All you have to do is:

1. download it
2. double-click it
3. add whatever you want
4. relax and let it manage your tasks for you!

Features:
- [x] Managing tasks (todos)
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon!)

If you're a Java programmer, you can use it to practice Java too! Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kheekheekhee committed:
Add .class to gitignore

kheekheekhee committed:
Finish level-1 with only basic functionalities like Greet, Echo and Exit

kheekheekhee committed:
Add ability to store user inputs and display them back to user and isolate greeting and exiting to their own functions

kheekheekhee committed:
Add ability to store text entered by user and display them back to user when requested and isolate greeting and exiting to their own functions

kheekheekhee committed:
Add the ability to mark tasks as done and create a Task class

kheekheekhee committed:
Add the ability to mark tasks as done and create a Task class

kheekheekhee committed:
Add support for the tasks ToDos, Events and Deadlines

kheekheekhee committed:
Add support for the tasks ToDos, Events and Deadlines

kheekheekhee committed:
Add text UI tests for functionality of greet, different tasks and exit

kheekheekhee committed:
Add text UI tests for functionality of greet, different tasks and exit

kheekheekhee committed:
Change the events test and actual outputs of events to "at" instead of "by"

kheekheekhee committed:
Change the events test and actual outputs of events to "at" instead of "by"

kheekheekhee committed:
Deal with errors such as incorrect inputs by user and add JavaDocs

kheekheekhee committed:
Deal with errors such as incorrect inputs by user and add JavaDocs

kheekheekhee committed:
Change typo in EmptyListException, add delete function and more tests to test exceptions and delete function

kheekheekhee committed:
Change typo in EmptyListException, add delete function and more tests to test exceptions and delete function

kheekheekhee committed:
Merge branch 'master' of https://github.com/kheekheekhee/ip

kheekheekhee committed:
Change access levels of attributes in Deadline, Event and Task to private

kheekheekhee committed:
Change access level of run method in Duke to private

kheekheekhee committed:
Fix some styling issues

kheekheekhee committed:
Add ability to save and load tasks to hard disk

kheekheekhee committed:
Change event and deadline to work with actual dates rather than strings

kheekheekhee committed:
Merge branch 'branch-Level-7'

kheekheekhee committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

kheekheekhee committed:
Refactor the code to extract out closely related code as classes

kheekheekhee committed:
Organize the classes into suitable java packages

kheekheekhee committed:
Add JUnit tests to test the behavior of the code

kheekheekhee committed:
Package the app as an executable JAR file

kheekheekhee committed:
Add JavaDoc comments to the code

kheekheekhee committed:
Update JavaDocs

kheekheekhee committed:
Fix and add comments

kheekheekhee committed:
Tweak the code the comply with coding standards

kheekheekhee committed:
Implement find feature

kheekheekhee committed:
Merge branch 'branch-Level-9'

kheekheekhee committed:
Merge branch 'branch-A-CodingStandard'

kheekheekhee committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/tasks/TaskList.java

kheekheekhee committed:
Add ability to print task lists in ui

kheekheekhee committed:
Update javadocs

kheekheekhee committed:
Update JavaDocs

kheekheekhee committed:
Update JavaDocs

kheekheekhee committed:
Update JavaDocs

kheekheekhee committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

kheekheekhee committed:
Merge branch 'branch-A-Gradle'

kheekheekhee committed:
Add checkstyle and update code with checkstyle suggestions

kheekheekhee committed:
Merge branch 'branch-A-CheckStyle'

kheekheekhee committed:
Add a GUI for Duke

kheekheekhee committed:
Merge branch 'branch-Level-10'

kevinchua6 commented on src/main/java/duke/exception/IncorrectFormatException.java:
```diff
@@ -0,0 +1,18 @@
+
+    /**
+     * Constructor of IncorrectFormatException.
+     * @param command event or deadline depending on what user called.
```
Should we have an empty line between the description and parameter section?

kevinchua6 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * Runs the command line version of the Duke chat bot.
+     */
+
```
Shouldn't we not have an empty line between the documentation block and the method?

kevinchua6 commented on src/main/java/duke/exception/InvalidIndexException.java:
```diff
@@ -0,0 +1,18 @@
+
+    /**
+     * Constructor of InvalidIndexException.
+     * @param start starting index of task list.
```
Should we have an empty line between the description and parameter section?

kevinchua6 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Handles and interprets user commands for Duke.
+ */
+
```
Maybe we should not have an empty line between the documentation block and the method to comply with the coding standard?

kevinchua6 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @param taskList contains all current tasks.
+     * @throws IOException if the named file exists but is a directory
+     * rather than a regular file, does not exist but cannot be created,
```
Should we indent continuation lines four (or more) spaces from the position of the `@`? I noticed the same issue in several other places too.

kevinchua6 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,247 @@
+
+    /**
+     * Adds a task to the list of Tasks with a confirmation message printed out after.
+     * @param task The duke.tasks.Task to be added to the list of Tasks
```
Should there be punctuation after each parameter description?

bryanwee023 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,96 @@
+
+        String input;
+
+        ui.printMessage(Ui.greet());
```
Maybe consider `ui.greet()` internally print the greeting message?

Rerouting ui's return value back to ui seems to break OOP's principle of encapsulation

bryanwee023 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * A launcher class to workaround classpath issues.
+ */
+public class Launcher {
+    public static void main(String[] args) {
+        Application.launch(Main.class, args);
+    }
+}
```
I'm not sure what's the purpose of Launcher. Is it a test class?

bryanwee023 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,147 @@
+                }
+                try {
+                    // excludes command "deadline " from the string
+                    return tasks.addDeadline(input.substring(9));
+                } catch (InvalidDateTimeException | MessageEmptyException | IncorrectFormatException e) {
+                    return e.getMessage();
+                }
+            case "todo":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // excludes command "todo" from the string
+                return tasks.addTodo(input.substring(5));
+            case "event":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // excludes command "event" from the string
+                return tasks.addEvent(input.substring(6));
+            case "delete":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // extracts index of task to delete
+                String deleteTaskIndex = words[words.length - 1];
+                return tasks.deleteTask(deleteTaskIndex);
+            case "find":
+                ArrayList<Task> matchedTasks = findCommand(words);
+                return Ui.printList(matchedTasks);
+            case "": // empty user input
+                throw new EmptyCommandException();
+            default: // all other inputs that are not supported
+                throw new InvalidCommandException();
+            }
+        } catch (DukeException | IOException e) {
+            return e.getMessage();
+        }
```
I think a switch-case flow might get messy if we have to continue adding more commands.
Maybe consider using a `Hashtable<String, Function>`? 

Alternatively, maybe handle error-checking within the methods invoked (eg. `tasks.deleteTask()`) instead of outside? I think it makes the code more readable because it trims the switch-case block and it complies to OOP's principle of encapsulation as well :)

bryanwee023 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,196 @@
+     * @return a new deadline that represents the deadline from the data file.
+     * @throws DataFileChangedException if the data file was changed and any entry contains a wrong format.
+     */
+    private Deadline extractDeadline(String text) throws DataFileChangedException {
```
Just curious to know. What's the reason for placing this function in the Storage class (rather than Deadline)?

bryanwee023 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,196 @@
+     * @throws DataFileChangedException if the data file was changed and any entry contains a wrong format.
+     */
+    private Event extractEvent(String text) throws DataFileChangedException {
+        int lastOccurrenceOfAt = text.lastIndexOf(" (at: ");
+        String description = text.substring(0, lastOccurrenceOfAt);
+
+        // disregards "( at: " and trailing ")"
+        String at = text.substring(lastOccurrenceOfAt + 6, text.length() - 1);
+
+        // prepare variables
+        String atWithoutWhiteSpace = at.replaceAll("\\s", "");
+        int lengthOfAtNoSpace = atWithoutWhiteSpace.length();
+
+        // throws error if it doesn't even contain sufficient number of characters for correct format
+        if (lengthOfAtNoSpace < 22 || lengthOfAtNoSpace > 25) { // MMM d yyyy, HH:mm - HH:mm
+            throw new DataFileChangedException();
+        }
```
It seems like the parsing of your input string takes up a chunk of your code. I think you can consider abstracting out all these details into a new StringParser class (or push it into the Task classes). 

Maybe doing so might make the code more reader-friendly?

bryanwee023 reviewed (commented):
Overall nice job! I like how your Event class has both a start and end time.

I think one thing that can be done is to abstract away details and push them into new methods or into the Task classes (which seems a little too empty)?

But other than that, the code is nicely commented and errors seem to be handled appropriately 👍

kheekheekhee committed:
Change code according to pull request comments

kheekheekhee committed:
Change code to follow suggestions from pull request comments

kheekheekhee committed:
Change code according to suggestions from pull request comments

kheekheekhee committed:
Change code to fit reommendations in pull request comments

kheekheekhee committed:
Change code to fit recommendations in pull request comments

kheekheekhee committed:
Change code to fit abstraction recommendations in pull request comments

kheekheekhee committed:
Change code to save to file after commands that change the list of tasks

kheekheekhee committed:
Add code for JavaFX window to close when bye command is called

kheekheekhee committed:
Add assertions to code

kheekheekhee committed:
Improve code quality

kheekheekhee committed:
Merge pull request #1 from kheekheekhee/branch-A-Assertions

Add assertions to code

kheekheekhee committed:
Merge branch 'master' into branch-A-CodeQuality

kheekheekhee committed:
Merge pull request #2 from kheekheekhee/branch-A-CodeQuality

Improve code quality

kheekheekhee committed:
Merge branch 'master' of https://github.com/kheekheekhee/ip

kheekheekhee committed:
Add data outputs to gitignore

kheekheekhee committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/kheekheekhee/ip into branch-A-CodeQuality

kheekheekhee committed:
Update gitignore

kheekheekhee committed:
Add ability to track expenses within tasks

kheekheekhee committed:
Update gitignore

kheekheekhee committed:
Merge branch 'branch-BCD-Extension'

# Conflicts:
#	.gitignore

kheekheekhee committed:
Fix some bugs

kheekheekhee committed:
Add ability to delete expenses and improve code for expense command

kheekheekhee committed:
Improve expense command code

kheekheekhee committed:
Add ability to sum expenses within a Task

Expenses within a Task cannot be summed up.

Calculating expenses without a summing function is very troublesome.

Let's implement the ability to detect and execute a /sum function.
It will find the chosen task, sum all expenses involved with that
Task and return the sum to the user.

Improves user experience as they do not need to sum up expenses
individually.

kheekheekhee committed:
Add ability to sum all expenses from all Tasks

There is no function to calculate the total expenses in the
TaskList and the user has to personally do it.

Summing all adds convenience to the user and is a typically
sought after function when expenses are involved.

Let's implement a function that makes use of the pre-existing
sumExpense function in each Task to sum total expenses in the
whole TaskList.

No Task and expense will be missed by looping through the whole
list and it adds a layer of convenience to the user.

kheekheekhee committed:
Fix bugs regarding ExpenseCommand

/sum, /sumall, /delete and /listall commands are case sensitive.
Error occurs when adding random messages before and after most
expense-related commands.

/sum, /sumall, /delete and /listall being case insensitive is more
user-friendly beacuse users cannot be expected to remember the
exact keywords. Better error checking for related commands are set
to prevent errors from incorrect inputs.

Let's,
* update the keyword-matching algorithm to use case-insensitive
matching
* add more error handling and stricter conditions for keyword to be
detected

Improves user experience greatly and expense-related commands
can function normally.

kheekheekhee committed:
Merge branch 'branch-A-CodeQuality'

kheekheekhee committed:
Add app icon and name to UI window

kheekheekhee committed:
Add user guide

kheekheekhee committed:
Update user guide

kheekheekhee committed:
Set theme jekyll-theme-tactile

kheekheekhee committed:
Add ability for expense to properly detect decimal points

kheekheekhee committed:
Merge branch 'master' of https://github.com/kheekheekhee/ip

kheekheekhee committed:
Improve JavaDocs

kheekheekhee committed:
Update formatting in docs/README.md

kheekheekhee committed:
Update README.md

kheekheekhee committed:
Update docs/README.md

kheekheekhee committed:
Merge branch 'master' of https://github.com/kheekheekhee/ip

kheekheekhee committed:
Update error messages for expense command
